presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java

```diff
@@ -39,7 +47,7 @@ protected void setup(Binder binder)
 
             binder.bind(PrestoS3FileSystemStats.class).toInstance(PrestoS3FileSystem.getFileSystemStats());
             newExporter(binder).export(PrestoS3FileSystemStats.class)
-                    .as(generator -> generator.generatedNameOf(PrestoS3FileSystem.class));
+                    .as(generator -> generator.generatedNameOf(PrestoS3FileSystem.class, catalogName));
```

Comment:
You can use provider of HiveCatalogName and pass function here to generate name lazy. After guice configuration

Reply:
I tried to replace `catalogName` with `binder.getProvider(HiveCatalogName.class).get().toString()` but I'm getting the following error:
```
2020-03-13T16:45:14.015+0100	ERROR	main	io.prestosql.server.PrestoServer	Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.NullPointerException: moduleSource cannot be null.
  at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:40)
  at org.weakref.jmx.guice.MBeanModule.configure(MBeanModule.java:39)
  while locating org.weakref.jmx.guice.GuiceMBeanExporter

1 error
com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.NullPointerException: moduleSource cannot be null.
  at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:40)
  at org.weakref.jmx.guice.MBeanModule.configure(MBeanModule.java:39)
  while locating org.weakref.jmx.guice.GuiceMBeanExporter

1 error
	at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:543)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:87)
	at io.airlift.bootstrap.Bootstrap.initialize(Bootstrap.java:245)
	at io.prestosql.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.prestosql.plugin.hive.HiveConnectorFactory.create(HiveConnectorFactory.java:52)
	at io.prestosql.connector.ConnectorManager.createConnector(ConnectorManager.java:341)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:203)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:195)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:181)
	at io.prestosql.metadata.StaticCatalogStore.loadCatalog(StaticCatalogStore.java:88)
	at io.prestosql.metadata.StaticCatalogStore.loadCatalogs(StaticCatalogStore.java:68)
	at io.prestosql.server.PrestoServer.run(PrestoServer.java:129)
	at io.prestosql.$gen.Presto_null____20200313_154403_1.run(Unknown Source)
	at io.prestosql.server.PrestoServer.main(PrestoServer.java:72)
Caused by: java.lang.NullPointerException: moduleSource cannot be null.
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:895)
	at com.google.inject.spi.ElementSource.<init>(ElementSource.java:94)
	at com.google.inject.spi.Elements$RecordingBinder.getElementSource(Elements.java:561)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:411)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:406)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:418)
	at io.prestosql.plugin.hive.s3.HiveS3Module.lambda$setup$0(HiveS3Module.java:43)
	at org.weakref.jmx.guice.Mapping.getName(Mapping.java:36)
	at org.weakref.jmx.guice.GuiceMBeanExporter.export(GuiceMBeanExporter.java:92)
	at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:42)
	at org.weakref.jmx.guice.GuiceMBeanExporter$$FastClassByGuice$$6cc63b28.newInstance(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
	at com.google.inject.internal.ConstructorInjector.access$000(ConstructorInjector.java:32)
	at com.google.inject.internal.ConstructorInjector$1.call(ConstructorInjector.java:98)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:112)
	at io.airlift.bootstrap.LifeCycleModule.provision(LifeCycleModule.java:54)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:120)
	at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:168)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:39)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:211)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:182)
	... 18 more
```
I'm open to suggestions, it seems the binder isn't ready to create a provider at this point (we're still during module setup).